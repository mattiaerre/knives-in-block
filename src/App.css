main {
  text-align: center;
}

section {
  border-bottom: 1px solid gray;
  margin-bottom: 40px;
}

ul {
  list-style-type: none;
  margin: 0;
  padding: 0;
}

.block,
.knives {
  /*
  background-color: red;
  */
}

.row {
  display: flex;
  justify-content: space-evenly;
}

.knife {
  cursor: pointer;
}

.disabled {
  cursor: default;
  opacity: 0.25;
}

.knife,
.slot {
  border: 1px solid gray;
  height: 20px;
  margin: 10px 40px;
  width: 40px;
}

.empty {
  background-color: black;
}

.blue {
  background-color: blue;
}

.turquoise {
  background-color: turquoise;
}

.white {
  background-color: white;
}

.debug {
  display: flex;
  flex-direction: column;
  text-align: left;
}
