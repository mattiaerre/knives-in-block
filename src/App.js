import './App.css';
import Knife from './components/Knife';
import Slot from './components/Slot';
import { useReducer } from 'react';

const initialArg = { block: null, knife: null };

function reducer(state, action) {
  switch (action.type) {
    case 'knife_pick_up': {
      return {
        ...state,
        knife: action.value
      };
    }
    case 'knife_put_down': {
      return {
        ...state,
        knife: null
      };
    }
    case 'slot_place_knife': {
      return {
        ...state,
        block: action.value,
        knife: null
      };
    }
    default:
      throw Error(`Unknown action: ${action.type}`);
  }
}

function App() {
  const [state, dispatch] = useReducer(reducer, initialArg);

  function getKnifeFromState() {
    return state.knife;
  }

  return (
    <main>
      <h1 className="source-sans-3-700">Jean Dubost</h1>
      <h2 className="source-sans-3-500">
        Laguiole Set of 6 Steak Knives in Block
      </h2>
      <section>
        <h3 className="source-sans-3-500">Knives</h3>
        TODO
      </section>
      <section>
        <h3 className="source-sans-3-500">Block</h3>
        TODO
      </section>
      <section>TODO</section>
      <section className="playground">
        <Knife color="blue" dispatch={dispatch} id="k1" />
        <Knife color="red" dispatch={dispatch} id="k2" />
        <Slot dispatch={dispatch} id="s1" onClick={getKnifeFromState} />
        <Slot dispatch={dispatch} id="s2" onClick={getKnifeFromState} />
      </section>
    </main>
  );
}

export default App;
