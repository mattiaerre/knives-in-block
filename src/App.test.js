import { fireEvent, render, screen } from '@testing-library/react';
import App from './App';

it('renders', () => {
  const { asFragment } = render(<App />);
  expect(asFragment()).toMatchSnapshot();
});

it('on click s1', () => {
  render(<App />);
  fireEvent.click(screen.getByTestId('s1'));
  expect(screen.getByTestId('s1')).toHaveClass('up');
});

it('on click s1 twice', () => {
  render(<App />);
  fireEvent.click(screen.getByTestId('s1'));
  fireEvent.click(screen.getByTestId('s1'));
  expect(screen.getByTestId('s1')).toHaveClass('down');
});

it('on click s7', () => {
  render(<App />);
  fireEvent.click(screen.getByTestId('s7'));
  expect(screen.getByTestId('s7')).toHaveClass('empty');
});

it('on click s1 then s2', () => {
  render(<App />);
  fireEvent.click(screen.getByTestId('s1'));
  fireEvent.click(screen.getByTestId('s2'));
  expect(screen.getByTestId('s1')).toHaveClass('up');
  expect(screen.getByTestId('s2')).toHaveClass('down');
});
