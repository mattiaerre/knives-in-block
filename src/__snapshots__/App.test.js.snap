// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders 1`] = `
<DocumentFragment>
  <main>
    <h1
      class="source-sans-3-700"
    >
      Jean Dubost
    </h1>
    <h2
      class="source-sans-3-500"
    >
      Laguiole Set of 6 Steak Knives in Block
    </h2>
    <section>
      <p
        style="text-align: right;"
      >
        <button
          class="source-sans-3-500"
        >
          Reset
        </button>
      </p>
    </section>
    <section>
      <h3
        class="source-sans-3-500"
      >
        Knives
      </h3>
      <ul
        class="knives"
      >
        <li>
          <div
            class="down slot"
            data-testid="s1"
            id="s1"
            style="background-color: blue;"
          />
        </li>
        <li>
          <div
            class="down slot"
            data-testid="s2"
            id="s2"
            style="background-color: blue;"
          />
        </li>
        <li>
          <div
            class="down slot"
            data-testid="s3"
            id="s3"
            style="background-color: turquoise;"
          />
        </li>
        <li>
          <div
            class="down slot"
            data-testid="s4"
            id="s4"
            style="background-color: turquoise;"
          />
        </li>
        <li>
          <div
            class="down slot"
            data-testid="s5"
            id="s5"
            style="background-color: white;"
          />
        </li>
        <li>
          <div
            class="down slot"
            data-testid="s6"
            id="s6"
            style="background-color: white;"
          />
        </li>
      </ul>
    </section>
    <section>
      <h3
        class="source-sans-3-500"
      >
        Block
      </h3>
      <ul
        class="block"
        style="border-color: lightgray;"
      >
        <li>
          <div
            class="empty slot"
            data-testid="s7"
            id="s7"
            style="background-color: black;"
          />
        </li>
        <li>
          <div
            class="empty slot"
            data-testid="s8"
            id="s8"
            style="background-color: black;"
          />
        </li>
        <li>
          <div
            class="empty slot"
            data-testid="s9"
            id="s9"
            style="background-color: black;"
          />
        </li>
        <li>
          <div
            class="empty slot"
            data-testid="s10"
            id="s10"
            style="background-color: black;"
          />
        </li>
        <li>
          <div
            class="empty slot"
            data-testid="s11"
            id="s11"
            style="background-color: black;"
          />
        </li>
        <li>
          <div
            class="empty slot"
            data-testid="s12"
            id="s12"
            style="background-color: black;"
          />
        </li>
      </ul>
    </section>
    <section>
      <h3
        class="source-sans-3-500"
      >
        Rules
      </h3>
      <p>
        Move knives into blocks to create beautiful patterns. Some are allowed, and the block border will turn 
        <span
          style="color: green; font-weight: bold;"
        >
          green
        </span>
        , while others are not, and the block border will turn 
        <span
          style="color: red; font-weight: bold;"
        >
          red
        </span>
        . As you place the knives, the block border will remain 
        <span
          style="color: lightgray; font-weight: bold;"
        >
          lightgray
        </span>
        .
      </p>
      <p>
        N.B.: Drag and drop doesn't work; try to click instead.
      </p>
    </section>
    <footer>
      <small
        class="source-sans-3-400"
      >
        knives-in-block v0.6.0 Â©2024
      </small>
    </footer>
  </main>
</DocumentFragment>
`;
